packages/gamut-labs/src/experimental/Popover/index.tsx

```diff
@@ -0,0 +1,156 @@
+   */
+  targetRef: React.RefObject<
+    Pick<HTMLDivElement, 'getBoundingClientRect' | 'contains'>
+  >;
+  /**
+   * Which vertical edge of the source component to align against.
+   * @default left
+   */
+  align?: 'left' | 'right';
+  /**
+   * Number of pixels to offset the popover from the source component.
+   * @default 20
+   */
+  offset?: number;
+  /**
+   * Which horizontal edge of the source componet to align against.
+   * @default below
+   */
+  position?: 'above' | 'below';
+  /**
+   * Whether to show a beak on the popover.
+   */
+  showBeak?: boolean;
+  /**
+   * Called when the Popover requests to be closed,
+   * this could be due to clicking outside of the popover, or by clicking the escape key.
+   */
+  onRequestClose?: () => void;
+  /**
+   * Called to prevent unwanted targetRef event (targetRef is outside of Popover).
+   * (i.e. when targetRef is a button and it is used to close the popover, this prevents triggering its onclick that would re-open the popover.)
+   */
+  disableOutsideEvent?: () => void;
+};
+
+export const Popover: React.FC<PopoverProps> = ({
+  className,
+  children,
+  isOpen,
+  showBeak,
+  position = 'below',
+  offset = 20,
+  align = 'left',
+  targetRef,
+  onRequestClose,
+  disableOutsideEvent,
+}) => {
+  const [targetRect, setTargetRect] = useState<DOMRect>();
+  const [isInViewport, setIsInViewport] = useState(true);
+  const { width, height } = useWindowSize();
+  const { x, y } = useWindowScroll();
+
+  const setRect = useCallback(() => {
+    const rect = targetRef?.current?.getBoundingClientRect();
+
+    setTargetRect(rect);
+  }, [targetRef]);
+
+  const isElementInViewPort = useCallback(() => {
+    if (targetRect) {
+      const inView =
+        targetRect.top >= 0 &&
+        targetRect.left >= 0 &&
+        targetRect.bottom <=
+          (window.innerHeight || document.documentElement.clientHeight) &&
+        targetRect.right <=
+          (window.innerWidth || document.documentElement.clientWidth);
+      setIsInViewport(inView);
+    }
+  }, [targetRect]);
+
+  const getPopoverPosition = useCallback(() => {
+    if (!targetRect) return {};
+
+    const positions = {
+      above: Math.round(window.scrollY + targetRect.top - offset),
+      below: Math.round(
+        window.scrollY + targetRect.top + targetRect.height + offset
+      ),
+    };
+    const alignments = {
+      right: Math.round(window.scrollX + targetRect.right),
+      left: Math.round(window.scrollX + targetRect.left),
+    };
+    return {
+      top: positions[position],
+      left: alignments[align],
+    };
+  }, [targetRect, offset, align, position]);
+
+  const handleClickOutside = (event: MouseEvent | KeyboardEvent) => {
+    if (event.type === 'keydown') {
+      onRequestClose?.();
+      return;
+    }
+    if (targetRef?.current?.contains(event.target as Element)) {
+      disableOutsideEvent?.();
+    }
+    onRequestClose?.();
```

Comment:
Nit: more ⛳ ! `onRequestClose?.();` happens either way, so the real difference is just the non-`'keydown'` case.

```suggestion
    if (event.type !== 'keydown' && targetRef?.current?.contains(event.target as Element)) {
      disableOutsideEvent?.();
    }
    onRequestClose?.();
```

_(that will probably make Prettier mad, but otherwise is nice IMO)_

Reply:
yes, I agree, this is much cleaner!